src/symbolize/gimli/libs_dl_iterate_phdr.rs

```diff
@@ -35,31 +35,53 @@ fn infer_current_exe(base_addr: usize) -> OsString {
     env::current_exe().map(|e| e.into()).unwrap_or_default()
 }
 
-// `info` should be a valid pointers.
-// `vec` should be a valid pointer to a `std::Vec`.
+/// # Safety
+/// `info` must be a valid pointer.
+/// `vec` must be a valid pointer to `Vec<Library>`
+#[forbid(unsafe_op_in_unsafe_fn)]
 unsafe extern "C" fn callback(
     info: *mut libc::dl_phdr_info,
     _size: libc::size_t,
     vec: *mut libc::c_void,
 ) -> libc::c_int {
-    let info = &*info;
-    let libs = &mut *vec.cast::<Vec<Library>>();
-    let is_main_prog = info.dlpi_name.is_null() || *info.dlpi_name == 0;
-    let name = if is_main_prog {
-        // The man page for dl_iterate_phdr says that the first object visited by
-        // callback is the main program; so the first time we encounter a
-        // nameless entry, we can assume its the main program and try to infer its path.
-        // After that, we cannot continue that assumption, and we use an empty string.
-        if libs.is_empty() {
-            infer_current_exe(info.dlpi_addr as usize)
-        } else {
+    // SAFETY: We are guaranteed these fields:
```

Comment:
Hm, the unsafe bit is dereferencing the `info` pointer which requires it to point to a valid `dl_phdr_info` struct (as defined by `libc`). Which, like `vec` being `Vec<Library>` is already part of the function's safety contract.

The fact that we're using only the guaranteed fields is good to know but that particular part could as well be safe code.

Reply:
What safe code are you imagining, exactly? This signature must conform to the callback signature defined by `dl_iterate_phdr`.

Reply:
The "these fields are guaranteed" comment is because the struct is **variably-sized**, and only those fields are guaranteed to be **initialized**.

Reply:
But the `*` dereferences the whole struct right? Which is the only unsafe part. The pattern matching is safe code after that.

Reply:
> dereferences

It creates a *place expression*.

But as I understand it, a pattern match only **loads** from the sites that it actually matches, thus, this:
```rust
let _ = unsafe { *info };
```
loads nothing and could actually be safe.

Reply:
I want to believe but is there an example I can run in miri to confirm this? My quick attempt failed but I don't know if it's really representative https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2c8fb800be3f138551cecda931fecec0

Reply:
hm, apparently it doesn't apply when pattern-matching, sigh.

Reply:
welp, loading four times instead i guess!